actsync/actionsync.go

```diff
@@ -0,0 +1,187 @@
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/iotexproject/iotex-core/pkg/fastrand"
+	"github.com/iotexproject/iotex-core/pkg/lifecycle"
+	"github.com/iotexproject/iotex-core/pkg/log"
+)
+
+const (
+	unicaseTimeout = time.Second
+	batchPeerSize  = 2
+)
+
+type (
+	// Neighbors acquires p2p neighbors in the network
+	Neighbors func() ([]peer.AddrInfo, error)
+	// UniCastOutbound sends a unicase message to the peer
+	UniCastOutbound func(context.Context, peer.AddrInfo, proto.Message) error
+
+	// ActionSync implements the action syncer
+	ActionSync struct {
+		lifecycle.Readiness
+		actions  sync.Map
+		syncChan chan hash.Hash256
+		wg       sync.WaitGroup
+		helper   *Helper
+		cfg      Config
+	}
+
+	actionMsg struct {
+		lastTime time.Time
+	}
+)
+
+// NewActionSync creates a new action syncer
+func NewActionSync(cfg Config, helper *Helper) *ActionSync {
+	return &ActionSync{
+		syncChan: make(chan hash.Hash256, cfg.Size),
+		helper:   helper,
+		cfg:      cfg,
+	}
+}
+
+// Start starts the action syncer
+func (as *ActionSync) Start(ctx context.Context) error {
+	as.wg.Add(1)
+	go as.sync()
+
+	as.wg.Add(1)
+	go as.triggerSync()
+
+	return as.TurnOn()
+}
+
+// Stop stops the action syncer
+func (as *ActionSync) Stop(ctx context.Context) error {
+	if err := as.TurnOff(); err != nil {
+		return err
+	}
+	close(as.syncChan)
+	as.wg.Wait()
+	return nil
+}
+
+// RequestAction requests an action by hash
+func (as *ActionSync) RequestAction(ctx context.Context, hash hash.Hash256) error {
+	if !as.IsReady() {
+		return nil
+	}
+	// check if the action is already requested
+	_, ok := as.actions.LoadOrStore(hash, &actionMsg{})
+	if ok {
+		log.L().Debug("Action already requested", log.Hex("hash", hash[:]))
+		return nil
+	}
+	as.trigger(hash)
+	return nil
+}
+
+// ReceiveAction receives an action
+func (as *ActionSync) ReceiveAction(ctx context.Context, hash hash.Hash256) {
+	if !as.IsReady() {
+		return
+	}
+	as.actions.Delete(hash)
+}
+
+func (as *ActionSync) sync() {
+	defer as.wg.Done()
+	for hash := range as.syncChan {
+		log.L().Debug("syncing action", log.Hex("hash", hash[:]))
+		ctx, cancel := context.WithTimeout(context.Background(), unicaseTimeout)
+		defer cancel()
+		msg, ok := as.actions.Load(hash)
+		if !ok {
+			log.L().Debug("action not requested or already received", log.Hex("hash", hash[:]))
+			continue
+		}
+		if time.Since(msg.(*actionMsg).lastTime) < as.cfg.Interval {
+			log.L().Debug("action is recently requested", log.Hex("hash", hash[:]))
+			continue
+		}
+		msg.(*actionMsg).lastTime = time.Now()
+		if err := as.requestFromNeighbors(ctx, hash); err != nil {
+			log.L().Warn("Failed to request action from neighbors", zap.Error(err))
+		}
+	}
+}
+
+func (as *ActionSync) triggerSync() {
+	defer as.wg.Done()
+	ticker := time.NewTicker(as.cfg.Interval)
+	defer ticker.Stop()
+	for {
+		if !as.IsReady() {
+			return
```

Comment:
is it possible it runs to here before `as.TurnOn()` is run, then will return?

Reply:
for {
   if !as.IsReady() {
       yield(100ms)
   }
}
for {
    // event loop
}

Reply:
as discussed offline. added a quit chan to handle this